src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -205,36 +239,79 @@ uct_cuda_ipc_open_memhandle_mempool(uct_cuda_ipc_rkey_t *key,
                                     CUdeviceptr *mapped_addr)
 {
     CUmemAccessDesc access_desc = {};
-    CUdeviceptr dptr;
-    ucs_status_t status;
+    ucs_status_t status         = UCS_OK;
+    CUmemoryPool *mpool;
+    khiter_t khiter;
+    int khret;
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportFromShareableHandle(
-                &key->ph.pool, (void *)&key->ph.handle.fabric_handle,
-                CU_MEM_HANDLE_TYPE_FABRIC, 0));
-    if (status != UCS_OK) {
-        return status;
-    }
+    pthread_rwlock_rdlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    khiter = kh_get(cuda_ipc_rem_mpool_cache,
+                    &uct_cuda_ipc_rem_mpool_cache.hash,
+                    (char*)&key->ph.handle.fabric_handle);
+    if (ucs_likely(khiter != kh_end(&uct_cuda_ipc_rem_mpool_cache.hash))) {
+        key->ph.pool = kh_val(&uct_cuda_ipc_rem_mpool_cache.hash, khiter);
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(&dptr,
-                key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
-    if (status != UCS_OK) {
-        return status;
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                    *key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+        goto err;
     }
 
-    status = uct_cuda_ipc_init_access_desc(&access_desc);
-    if (status != UCS_OK) {
-        return status;
+    /* acquire writer lock as we did not find cached mpool in hashmap */
+    pthread_rwlock_unlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    pthread_rwlock_wrlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+
+    khiter = kh_put(cuda_ipc_rem_mpool_cache,
+                    &uct_cuda_ipc_rem_mpool_cache.hash,
+                    ucs_strndup((char*)&key->ph.handle.fabric_handle,
```

Comment:
Is it safe in terms of a memory leak? What is `kh_put` fails?

Reply:
Fixed this.